File: 2_medium/00200_number_of_islands/dfs.rs
impl Solution {
    fn dfs(grid: &mut Vec<Vec<char>>, directions: &[(i32, i32)], i: usize, j: usize) {

        if grid[i][j] == '0' {
            return;
        }

        grid[i][j] = '0';

        for &(x, y) in directions.iter() {

            let new_i_i32 = i as i32 + x;
            let new_j_i32 = j as i32 + y;

            if new_i_i32 < 0 || new_j_i32 < 0 {
                continue;
            }

            let new_i = new_i_i32 as usize;
            let new_j = new_j_i32 as usize;

            if new_i >= grid.len() || new_j >= grid[0].len() {
                continue;
            }

            Self::dfs(grid, directions, new_i, new_j);
        }
    }
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {

        let directions = [(0, 1), (0, -1), (1, 0), (-1, 0)];
        let mut count = 0_i32;
        let rows = grid.len();
        let cols = grid[0].len();

        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] == '1' {
                    count += 1;
                    Self::dfs(&mut grid, &directions, i, j);
                }
            }
        }

        count
    }
}

File: 2_medium/0001_1000/00200_number_of_islands/dfs.cpp
class Solution {
     void dfs(vector<vector<char>> &grid, vector<vector<int>> &directions, int i, int j) {

        if (grid[i][j] == '0') {
            return;
        }

        grid[i][j] = '0';

        for (auto &dir: directions) {

            int new_i = i + dir[0];
            int new_j = j + dir[1];

            if (new_i < 0 || new_j < 0 || new_i >= grid.size() || new_j >= grid[0].size()) {
                continue;
            }

            dfs(grid, directions, new_i, new_j);
        }
    }

public:
    int numIslands(vector<vector<char>>& grid) {
        vector<vector<int>> directions{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        int count = 0;
        int rows = grid.size();
        int cols = grid[0].size();

        for (int i{0}; i < rows; ++i) {
            for (int j{0}; j < cols; ++j) {
                if (grid[i][j] == '1') {
                    count += 1;
                    dfs(grid, directions, i, j);
                }
            }
        }

        return count;
    }
};
